src/librustdoc/passes/collect_trait_impls.rs

```diff
@@ -34,11 +34,12 @@ crate fn collect_trait_impls(mut krate: Crate, cx: &mut DocContext<'_>) -> Crate
 
     let mut new_items = Vec::new();
 
-    for &cnum in cx.tcx.crates(()).iter() {
-        for &(did, _) in cx.tcx.all_trait_implementations(cnum).iter() {
-            inline::build_impl(cx, None, did, None, &mut new_items);
-        }
+    // External trait impls.
+    let all_trait_impls = std::mem::take(&mut cx.resolver_caches.all_trait_impls);
+    for &impl_def_id in all_trait_impls.iter().skip_while(|def_id| def_id.is_local()) {
+        inline::build_impl(cx, None, impl_def_id, None, &mut new_items);
     }
+    cx.resolver_caches.all_trait_impls = all_trait_impls;
```

Comment:
This "move out and move back" thing is annoying.
`cx` (doc context) is mutably borrowed by any non-trivial logic, so you cannot put that logic into a loop that iterates over something inside `cx`.

I can add a helper method hiding these moves, if necessary.

Also, simply cloning `all_traits`/`all_trait_impls` in cases like this is always an alternative, it should even be cheap compared to the rest of the logic here, it just feels like a surrender.

Reply:
Personally, I like the idea of a helper, but have no strong opinion. Only thought with a helper is that it may avoid needing to remember to move back afterwards.

Reply:
Added helpers and also added checks ensuring that we are not trying to move the caches for the second time.